main/src/ca/uwaterloo/flix/language/phase/unification/set/SetUnification.scala

```diff
@@ -17,29 +17,284 @@
+
+  object Options {
+    sealed trait Debugging
+
+    object Debugging {
+      case object Nothing extends Debugging
+
+      case object All extends Debugging
+    }
+  }
+
+  /**
+    * Represents the running state of the solver.
+    *   - `eqs`: the remaining equations to solve
+    *   - `errEqs`: the remaining equations that have been found unsolvable
+    *   - `subst`: the current substitution, which has already been applied to `eqs`
+    *   - `lastProgressPhase`: the name of the last phase that was run and made progress
+    *   - `currentPhase`: the number of the last phase that was run
+    */
+  private class State(var eqs: List[Equation]) {
+    var errEqs: List[Equation] = Nil
+    var subst: SetSubstitution = SetSubstitution.empty
+    var lastProgressPhaseName: String = "Setup (Nothing)"
+    var lastProgressPhaseNumber: Int = 0
+    var currentPhaseNumber: Int = 0
+
+    def allEqs: List[Equation] = eqs ++ errEqs
+  }
+
+  /**
+    * Moves all non-pending equations into their own list for performance, not
+    * required for correctness.
+    * */
+  private def moveErrEquations(state: State): Unit = {
+    state.eqs = state.eqs.filter(eq => {
+      if (eq.isPending) {
+        true
+      } else {
+        state.errEqs = eq :: state.errEqs
+        false
+      }
+    })
+  }
+
+  /**
+    * Attempts to solve the equation system `eqs`.
+    *
+    * All returned equations are either marked [[Equation.Status.Unsolvable]] or
+    * [[Equation.Status.Timeout]].
+    */
+  def solve(eqs: List[Equation])(implicit opts: Options): (List[Equation], SetSubstitution) = {
+    val (solvedEqs, s, _) = solveWithInfo(eqs)
+    (solvedEqs, s)
+  }
+
+  /**
+    * Attempts to solve the equation system `eqs`, like [[solve]].
+    *
+    * Additionally returns the name and number of the last phase to make progress.
     */
-  private final case class Options(sveRecSizeThreshold: Int)
+  def solveWithInfo(l: List[Equation])(implicit opts: Options): (List[Equation], SetSubstitution, (String, Int)) = {
+    val noDebug = opts.copy(debugging = Options.Debugging.Nothing)
 
-  /** A class to track progress during set unification. */
-  private final class Progress() {
+    val state = new State(l)
 
-    /** Is `true` if [[markProgress]] has been called. */
-    private var hasMadeProgress: Boolean = false
+    runPhase(runRule(constantAssignment), state, "Constant Assignment", "Solves equations where one or both sides are ground.")
+    runPhase(runRule(trivial), state, "Trivial Equations", "Solves trivial equations.")(noDebug)
+    runPhase(runRule(variableAlias), state, "Variable Aliases", "Solves equations like `x1 ~ x2`.")
+    runPhase(runRule(trivial), state, "Trivial Equations", "Solves trivial equations.")(noDebug)
+    runPhase(runRule(variableAssignment), state, "Simple Variable Assignment", "Solves non-recursive variable assignments.")
+    runPhase(runRule(trivial), state, "Trivial Equations", "Solves trivial equations.")(noDebug)
```

Comment:
Perhaps we need a comment at the top that the equations are passed implicitly between the phase?

Reply:
Like in Flix.scala or Matt's solver the asts/equations are actually explicitly threaded through.
So you can breakpoint on a phase.

Reply:
Hmm yeah. State is passed explitictly so its not that hidden.

I'm not sure how I'd do the state update separately (inline `runPhase`), that'd be very verbose

Reply:
You can breakpoint inside `runPhase` to get the same effect, although you need to skip through the couple you don't care about.

Reply:
We can't pipeline it?

Reply:
I guess each `runWithState(..)` becomes
```
debugPhase(state, "Sve", "..")
val rule5 = runRule(..)(output4)
val output5 = updateState(rule5)
```

Reply:
hmm but also breaks the tracking of the last phase to make progress

Reply:
I think the flexibiliy of runWithState is better. I feel the upsides outweigh the ability to put a breakpoint exactly on the phase